### PR TITLE
tools: add a tool to do sanity check for topology and firmware

### PR DESCRIPTION
**Pull request by RanderWang**
With the rise of closed source and a large number of processing components, we could sanity check the topology against the firmware rather than having a cryptic IPC failure when the topology requests a component not in the firmware.

usage:  **tplgfw_check.py [-h] [--version] -t T [T ...] -f F [F ...] [-v]** 
_tplgfw_check.py -t sof-tgl-nocodec.tplg -f basefw.bin_

It is for **[FEATURE] Allow topologies to manifest and sanity check component IDs** https://github.com/thesofproject/sof/issues/3871

Validation case:
**case 1:  Ref FW + ipc4 topology without smart amp and eqiir pass**
rander@rander-i9:~/sof/sof/tools/sof_tplg_fw_sanity_check$ ./tplgfw_check.py -f /home/rander/dsp_fw_mtl_release.bin -t /home/rander/sof-mtl-nocodec.tplg 
FW header Container: 
    id = u'$AE1' (total 4)
    len = 2260
    version_major = 1
    version_minor = 0
    module_entries = 27 

IPC4 supported FW

This FW can support with this topology

**case 2:  Ref FW + ipc4 topology with smart amp and eqiir failed**
rander@rander-i9:~/sof/sof/tools/sof_tplg_fw_sanity_check$ ./tplgfw_check.py -f /home/rander/dsp_fw_mtl_release.bin -t /home/rander/sof/sof/tools/build_tools/topology/topology2/development/sof-mtl-nocodec.tplg 
FW header Container: 
    id = u'$AE1' (total 4)
    len = 2260
    version_major = 1
    version_minor = 0
    module_entries = 27 

IPC4 supported FW

Error: widget: smart_amp.2.1 b'\x1e\x96z\x16\xe4\x8a\xea\x11\x89\xf1\x00\x0c)\xce\x165' is not supported by this FW
Error: widget: eqiir.14.1 b"\xe6\xc0PQ\xf9'\xc8N\x83Q\xc7\x05\xb6B\xd1/" is not supported by this FW

**case 3:  IPC4 cSOF FW + ipc4 topology pass**
rander@rander-i9:~/sof/sof/tools/sof_tplg_fw_sanity_check$ ./tplgfw_check.py -f /home/rander/zephyr/build-sof-staging/sof/community/sof-mtl.ri -t /home/rander/sof/sof/tools/build_tools/topology/topology2/development/sof-mtl-nocodec.tplg 
FW header Container: 
    id = u'$AE1' (total 4)
    len = 1948
    version_major = 1
    version_minor = 0
    module_entries = 19 

IPC4 supported FW

This FW can support with this topology

rander@rander-i9:~/sof/sof/tools/sof_tplg_fw_sanity_check$ ./tplgfw_check.py -f /home/rander/zephyr/build-sof-staging/sof/dbgkey/sof-tgl.ri -t /home/rander/sof/sof/tools/build_tools/topology/topology2/development/sof-tgl-nocodec.tplg 
FW header Container: 
    id = u'$AE1' (total 4)
    len = 1580
    version_major = 1
    version_minor = 0
    module_entries = 15 

IPC4 supported FW

This FW can support with this topology

**case 4:  IPC4 cSOF FW + ipc3 topology failed**
rander@rander-i9:~/sof/sof/tools/sof_tplg_fw_sanity_check$ ./tplgfw_check.py -f /home/rander/zephyr/build-sof-staging/sof/dbgkey/sof-tgl.ri -t /home/rander/sof/sof_tmp/tools/build_tools/topology/topology1/development/sof-tgl-nocodec-ci.tplg 
FW header Container: 
    id = u'$AE1' (total 4)
    len = 1580
    version_major = 1
    version_minor = 0
    module_entries = 15 

IPC4 supported FW

Error: widget: PCM0P b'\x0c\x10\x9d\x8bxm\x8fA\x90\xa3\xe0\xe8\x05\xd0\x85+' is not supported by this FW
Error: widget: PGA1.1 b'~g~\xb7\xf4_\x88A\xaf\x14\xfb\xa8\xbd\xbf\x86\x82' is not supported by this FW
Error: widget: MUXDEMUX1.0 b'hh\xb2\xc40\x14\x0eG\xa0\x89\x15\xd1\xc7\x7f\x85\x1a' is not supported by this FW
......
Error: widget: DMIC0.IN b"'\r\xb0\xc2\xbc\xffPA\xa5\x1a$\\y\xc5\xe5K" is not supported by this FW
Error: widget: PCM5C b'\x0c\x10\x9d\x8bxm\x8fA\x90\xa3\xe0\xe8\x05\xd0\x85+' is not supported by this FW
Error: widget: DMIC1.IN b"'\r\xb0\xc2\xbc\xffPA\xa5\x1a$\\y\xc5\xe5K" is not supported by this FW

**case 5:  IPC3 SOF FW + ipc4 topology failed**
rander@rander-i9:~/sof/sof/tools/sof_tplg_fw_sanity_check$ ./tplgfw_check.py -f /home/rander/Downloads/sof-tgl.ri -t /home/rander/sof/sof/tools/build_tools/topology/topology2/development/sof-tgl-nocodec.tplg 
FW header Container: 
    id = u'XMan' (total 4)
    len = 736
    version_major = 16
    version_minor = 0
    module_entries = 16777216 

IPC3 supported FW

Error: widget: copier.host.1.1 b'\x83\x0c\xa0\x9b\x12\xca\x83J\x94<\x1f\xa2\xe8/\x9d\xda' is not supported by this FW
Error: widget: gain.1.1 b'\xa8\xa9\xbca\xd0\x18\x18J\x8e{&9!\x98\x04\xb7' is not supported by this FW
Error: widget: mixin.1.1 b'\xb2ne9q;I@\x8d?\xf9,\xd5\xc4<\t' is not supported by this FW
......
Error: widget: copier.host.21.1 b'\x83\x0c\xa0\x9b\x12\xca\x83J\x94<\x1f\xa2\xe8/\x9d\xda' is not supported by this FW
Error: widget: gain.21.1 b'\xa8\xa9\xbca\xd0\x18\x18J\x8e{&9!\x98\x04\xb7' is not supported by this FW
Error: widget: mixin.21.1 b'\xb2ne9q;I@\x8d?\xf9,\xd5\xc4<\t' is not supported by this FW

### CODE REVIEW

File: tools/sof_tplg_fw_sanity_check/tplgfw_check.py [hunk 1]
```diff
@@ -0,0 +1,739 @@
+#!/usr/bin/env python3
```
Comment (kv2019i) [inline tools/sof_tplg_fw_sanity_check/tplgfw_check.py:1]: I'd maybe reword "rise of closed source" and just have "With increasing number of processing components...".

File: tools/sof_tplg_fw_sanity_check/tplgfw_check.py [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+                        if cmd_args.verbose:
+                            print("Widget: ", name, "is supported by ", m.name)
+                        break;
+
+                if found == False:
+                    not_found += 1
+                    print("Error: widget:", name, uuid, "is not supported by this FW")
+
+        if not_found == 0:
+            print("This FW can support with this topology")
```
Comment (kv2019i) [inline tools/sof_tplg_fw_sanity_check/tplgfw_check.py:737]: maybe: "FW is compatible with the topology."

please also return 0 in case of success and a negative error in case the test fails. then this script can be used as tool by other app/scripts.

**Comment by ujfalusi** [comment]
@RanderWang, I'm not sure what is cryptic about this message:
```
Direct firmware load for intel/avs-lib/tgl/community/5150C0E6-27F9-4EC8-8351-C705B642D12F.bin failed with error -2
Library file 'intel/avs-lib/tgl/community/5150C0E6-27F9-4EC8-8351-C705B642D12F.bin' is missing
failed to find module info for widget eqiir.4.1 with UUID 5150C0E6-27F9-4EC8-8351-C705B642D12F
```
We cannot find the module (not in basefw and not available as loadable) for `eqiir.4.1` which has UUID `5150C0E6-27F9-4EC8-8351-C705B642D12F`

On the other hand:
```
Error: widget: eqiir.14.1 b"\xe6\xc0PQ\xf9'\xc8N\x83Q\xc7\x05\xb6B\xd1/" is not supported by this FW
```
One would have to decipher what that `"\xe6\xc0PQ\xf9'\xc8N\x83Q\xc7\x05\xb6B\xd1/" ` is, how that is an UUID.

Does this tool checks for external libraries? Closed source might rise outside of open source ;)

**Review by kv2019i — APPROVED** [review]
That's a cool tool and a welcome addition! I'd run this through pylint (if you haven't already). 

I'm guessing we don't have other Python tools yet to parse the tplg (we have the tplg_parser in C). But I could imagine this could be used as basis for other tools as well.

File: tools/sof_tplg_fw_sanity_check/tplgfw_check.py (hunk 1 above)
Comment (cujomalainey) [reply tools/sof_tplg_fw_sanity_check/tplgfw_check.py:1]: To be honest though it sounds like a great start to a horror tale for the linux mailing list

**Review by cujomalainey — COMMENTED** [review]
great tool, only thing is python in my opinion is not great at handling raw bytes. Might be worth checking out this which lets you just build it into the type then pretty print the debug and it does all of its range checks on enums for you https://github.com/cujomalainey/ant-rs/blob/development/ant/src/messages/config.rs#L43

**Review by marc-hb — CHANGES_REQUESTED** [review]
This is a MASSIVE copy/paste/diverge of https://github.com/thesofproject/sof-test/blob/main/tools/tplgtool2.py without even pointing it.

This is python and 2023, there is no reason for so much copy/paste/diverge. Don't do this.

The obvious fix is to convert tplgtool2.py to a library that you can `import` and re-use
https://realpython.com/python-main-function/#a-basic-python-main

> great tool, only thing is python in my opinion is not great at handling raw bytes.

Python alone is not but `construct` used here is pretty good. Also, Python is familiar to many more people (at least for now) and I don't think Rust's low-level performance and memory management are needed here.

**Review by fredoh9 — CHANGES_REQUESTED** [review]
Instead of copying tplgtool2.py here, can you submit a patch in sof-test and use it from sof-test?

**Comment by marc-hb** [comment]
> Instead of copying tplgtool2.py here, can you submit a patch in sof-test and use it from sof-test?

I was about to suggest more or less the same thing. Why can't `sof-test/tplgtool2.py` do both? 1. What it already does now, and 2. this new thing.

**Comment by lgirdwood** [comment]
> > Instead of copying tplgtool2.py here, can you submit a patch in sof-test and use it from sof-test?
> 
> I was about to suggest more or less the same thing. Why can't `sof-test/tplgtool2.py` do both? 1. What it already does now, and 2. this new thing.

Lets move all sof-test stress/smoke/validation tests that are useful for developers to the main sof repo. e.g. to ```sof/test``` directory in the main repo. Having a separate test repo today gives the impression that its for testers only, testing is for developers so lets make it more obvious. CI/validation only tests should remain in sof-test if we have any.

**Comment by plbossart** [comment]
> > > Instead of copying tplgtool2.py here, can you submit a patch in sof-test and use it from sof-test?
> > 
> > 
> > I was about to suggest more or less the same thing. Why can't `sof-test/tplgtool2.py` do both? 1. What it already does now, and 2. this new thing.
> 
> Lets move all sof-test stress/smoke/validation tests that are useful for developers to the main sof repo. e.g. to `sof/test` directory in the main repo. Having a separate test repo today gives the impression that its for testers only, testing is for developers so lets make it more obvious. CI/validation only tests should remain in sof-test if we have any.

@lgirdwood it's not really practical to move everything under the sof/ umbrella. sof-test has different reviewers, should not be tied to releases, has very strong ties to CI, and you would create mutual dependencies between PRs for sof and sof-test within sof.

**Comment by marc-hb** [comment]
> sof-test [...] should not be tied to releases, has very strong ties to CI, and you would create mutual dependencies between PRs for sof and sof-test within sof.

This would indeed make testing and reproducing failures in SOF branches much more complicated. Imagine if SOF git bisect moved the test code at the same time, good luck with that! The only workaround would probably be to git clone sof twice!! Once for the firmware and another clone for the tests. Then most users including CI would probably name the second SOF clone... "sof-test" :-)

A more accurate name for `sof-test` would be `ALSA-and-topologies-test-suite` and ALSA lifecycle's is independent from SOF. Unlike SOF unit tests, the coupling between sof and sof-test is very loose: the Linux kernel and ALSA sit between them.

sof-ci/jenkins has a strong dependency on sof-test but it's only one-way: sof-test has zero dependency on Intel CI. In fact sof-test had contributions from non-Intel Yocto users.

> Lets move all sof-test stress/smoke/validation tests that are useful for developers [...] CI/validation only tests should remain in sof-test if we have any.

I don't know what is a "validation-only" test. Did you mean Intel test _plans_ maybe? There is no test plan in sof-test right now; 100% of the content in sof-test can (and must) be used by all developers using ALSA right now. If you're not already using sof-test then you're not an SOF developer, I mean you're not an SOF developer using Linux and ALSA

> Having a separate test repo today gives the impression that its for testers only, testing is for developers so lets make it more obvious.

If one additional `git clone` command or a mere repo name is all it takes to discourage some developer from reproducing a test failure then... I'm not impressed. But maybe we can help even such test-allergic developer(s) with this simple change:

```diff
--- sof/west.yml
+++ sof/west.yml
@@ -59,6 +59,11 @@ manifest:
           - lz4
           - tinycrypt
 
+    # Linux ALSA and topologies test suite
+    - name: sof-test
+      repo-path: sof-test
+      revision: main
+
   self:
     # Changes to submanifests/*.yml files _are_ effective; these have no
     # specified 'revision:'
```

That's probably where most Linux people have sof-test already cloned anyway.

It would be pointless for Windows developers but it's small.

It is typical to find `west update` in an SOF `git bisect` script, this would have to change to something like `west update zephyr rimage tomlc99` to avoid moving sof-test, maybe not a bad thing.

**Comment by ranj063** [comment]
> @RanderWang, I'm not sure what is cryptic about this message:
> 
> ```
> Direct firmware load for intel/avs-lib/tgl/community/5150C0E6-27F9-4EC8-8351-C705B642D12F.bin failed with error -2
> Library file 'intel/avs-lib/tgl/community/5150C0E6-27F9-4EC8-8351-C705B642D12F.bin' is missing
> failed to find module info for widget eqiir.4.1 with UUID 5150C0E6-27F9-4EC8-8351-C705B642D12F
> ```
> 
> We cannot find the module (not in basefw and not available as loadable) for `eqiir.4.1` which has UUID `5150C0E6-27F9-4EC8-8351-C705B642D12F`
> 
> On the other hand:
> 
> ```
> Error: widget: eqiir.14.1 b"\xe6\xc0PQ\xf9'\xc8N\x83Q\xc7\x05\xb6B\xd1/" is not supported by this FW
> ```
> 
> One would have to decipher what that `b"\xe6\xc0PQ\xf9'\xc8N\x83Q\xc7\x05\xb6B\xd1/"` is, how that is an UUID.
> 
> Does this tool checks for external libraries? Closed source might rise outside of open source ;)

yeah seriously why do we need a new tool for this? The kernel error message is quite self-explanatory that the base FW does not support the module or there's no external library available for it either. I don't see what additional value this tool brings on top what the kernel dmesg already tells us.

**Comment by marc-hb** [comment]
>  I'm not sure what is cryptic about this message:

Error messages can be fixed, this is a digression.

The relevant question is: "Is it valuable to parse a topology file without having to 1. find the corresponding hardware 2. boot a linux kernel or restart the firmware". I don't have the answer but I imagine it could be useful for some larger scale validation?

Another cool feature of `construct` is the ability to edit binaries interactively from the Python shell:
https://construct.readthedocs.io/en/latest/intro.html#example

PS: if this had been submitted as a 10 times smaller patch in sof-test then most people here would have not noticed. Maybe it's not too late :-)

**Comment by plbossart** [comment]
I think there's value in checking that the topology makes references to valid UUIDs.

But this is something that the kernel should do prior to initiating any IPCs.
The fact that we get a cryptic IPC fails is not really problematic IMHO, what's problematic is that we do not detect this problem earlier by cross-checking what the topology wants to use and what the firmware + modules can provide.

btw, 'with the rise of modules', it's impossible to statically check if an UUID is supported or not. The UUID may well well be supported by a module that is different from the base firmware. 

"tplgfw_check.py -t sof-tgl-nocodec.tplg -f basefw.bin" is completely useless IMHO, one would need the list of all additional modules as well.

**Comment by ranj063** [comment]
> I think there's value in checking that the topology makes references to valid UUIDs.
> 
> But this is something that the kernel should do prior to initiating any IPCs. The fact that we get a cryptic IPC fails is not really problematic IMHO, what's problematic is that we do not detect this problem earlier by cross-checking what the topology wants to use and what the firmware + modules can provide.
@plbossart this is exactly what we do. The kernel checks for all necessary module info during topology parsing and if not found, the audio card wont even be registered.

**Comment by plbossart** [comment]
> > I think there's value in checking that the topology makes references to valid UUIDs.
> > But this is something that the kernel should do prior to initiating any IPCs. The fact that we get a cryptic IPC fails is not really problematic IMHO, what's problematic is that we do not detect this problem earlier by cross-checking what the topology wants to use and what the firmware + modules can provide.
> @plbossart this is exactly what we do. The kernel checks for all necessary module info during topology parsing and if not found, the audio card wont even be registered.

that's not really what the log above shows, we attempt to load a module and then fail.

**Comment by ranj063** [comment]
> that's not really what the log above shows, we attempt to load a module and then fail.

No @plbossart . The error message is from https://github.com/thesofproject/linux/blob/1a7de96f85e37b4f879f8ef6d15c337959336c93/sound/soc/sof/ipc4-topology.c#L329.

This function is called during sof_topology_complete() when we make sure that the IPC structures for all widgets in topology are setup

**Comment by RanderWang** [comment]
@plbossart  @ranj063 I may misunderstand this feature : [FEATURE] Allow topologies to manifest and sanity check component IDs https://github.com/thesofproject/sof/issues/3871. In my opinion we need a tool to do static check without driver. If our driver can check these UUID before IPC communication, then we don't need this tool and close the feature requirement ?

**Comment by RanderWang** [comment]
Thanks, I will add this feature to tplgtool2.py directly and close this PR.
